mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,279 @@ impl MixnetNode {
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
-            .await
-            .map_err(Into::into);
-        }
+
+        self.message_tx.send(TargetedMessage::new(addr, msg))?;
         Ok(())
     }
 }
+
+struct MessageHandler {
+    // TODO: remove this allow when we implement the retry logic
+    #[allow(dead_code)]
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+    message_tx: mpsc::UnboundedSender<TargetedMessage>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl MessageHandler {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<TargetedMessage>,
+        message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        match self.send(msg).await {
+                            Ok(Some(msg)) => {
+                                let _ = self.message_tx.send(msg);
+                            }
+                            Ok(None) => {},
+                            Err(e) => {
+                                tracing::error!("failed to send msg: {e}");
+                            },
+                        }
+                    } else {
+                        // Channel closed, we should shutdown the message handler thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down message handler thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    /// Send a message to the remote node,
+    /// return Ok(Some((Duration, Message))) if the message is not sent and the error is retryable
+    /// return Ok(None) if the message is sent successfully
+    /// return Err(e) if the message is not sent and the error is not retryable
+    async fn send(&mut self, mut msg: TargetedMessage) -> Result<Option<TargetedMessage>> {
```

Comment:
This is hard to read even with a doc comment. I would suggest we create a new enum with more descriptive variantas